phoenix-core/src/it/java/org/apache/phoenix/end2end/index/MutableIndexFailureIT.java

```diff
@@ -135,230 +154,282 @@ public void tearDown() throws Exception {
-              rs = conn.getMetaData().getTables(null, StringUtil.escapeLike(SCHEMA_NAME), INDEX_TABLE_NAME+"_2",
+            if(localIndex) {
+                rs = conn.getMetaData().getTables(null, StringUtil.escapeLike(SCHEMA_NAME), INDEX_TABLE_NAME+"_2",
+                    new String[] { PTableType.INDEX.toString() });
+                assertTrue(rs.next());
+                assertEquals(INDEX_TABLE_NAME+"_2", rs.getString(3));
+                assertEquals(PIndexState.DISABLE.toString(), rs.getString("INDEX_STATE"));
+                assertFalse(rs.next());
+            }
+    
+            // if the table is transactional the write to the index table will fail because the
+            // index has not been disabled
+            if (!transactional) {
+                // Verify UPSERT on data table still work after index is disabled
+                stmt = conn.prepareStatement("UPSERT INTO " + DATA_TABLE_FULL_NAME + " VALUES(?,?,?)");
+                stmt.setString(1, "a3");
+                stmt.setString(2, "x3");
+                stmt.setString(3, "3");
+                stmt.execute();
+                conn.commit();
+            }
+            
+            if (transactional) {
+                // if the table was transactional there should be 1 row (written before the index
+                // was disabled)
+                query = "SELECT /*+ NO_INDEX */ v2 FROM " + DATA_TABLE_FULL_NAME;
+                rs = conn.createStatement().executeQuery("EXPLAIN " + query);
+                String expectedPlan =
+                        "CLIENT PARALLEL 1-WAY FULL SCAN OVER " + DATA_TABLE_FULL_NAME;
+                assertEquals(expectedPlan, QueryUtil.getExplainPlan(rs));
+                rs = conn.createStatement().executeQuery(query);
+                assertTrue(rs.next());
+                assertEquals("1", rs.getString(1));
+                assertFalse(rs.next());
+            } else {
+                // if the table was not transactional there should be three rows (all writes to data
+                // table should succeed)
+                query = "SELECT v2 FROM " + DATA_TABLE_FULL_NAME;
+                rs = conn.createStatement().executeQuery("EXPLAIN " + query);
+                String expectedPlan =
+                        "CLIENT PARALLEL 1-WAY FULL SCAN OVER " + DATA_TABLE_FULL_NAME;
+                assertEquals(expectedPlan, QueryUtil.getExplainPlan(rs));
+                rs = conn.createStatement().executeQuery(query);
+                assertTrue(rs.next());
+                assertEquals("1", rs.getString(1));
+                assertTrue(rs.next());
+                assertEquals("2", rs.getString(1));
+                assertTrue(rs.next());
+                assertEquals("3", rs.getString(1));
+                assertFalse(rs.next());
+            }
+            
+            // recreate index table
+            admin.createTable(indexTableDesc);
+            do {
+              Thread.sleep(15 * 1000); // sleep 15 secs
```

Comment:
How is this sleep helping? Is it account for the async nature of the API? FWIW, the admin.createTable() already retries before it throws TableNotFoundException. If the number of retries is too small then you could possibly increasing the number of retries too.